### ISSUE TITLE
<amp-script>: should AMP.setState() return a Promise?

### ISSUE THREAD

**Issue by morsssss** [ampproject/worker-dom#806]
In `<amp-script>`, `AMP.getState()` returns a Promise. But `AMP.setState()` doesn't. This... is confusing, since I'd assume that both of these depend on sending a message to the main thread. In either case, I think it's confusing that the interface is different for two very similar methods.

I might think it would be simplest if both `getState()` and `setState()` were simply synchronous. I'd guess there aren't a lot of use cases where your script would want to request the contents of a state variable, then go on with other activities until the contents return. Even if this makes any sense, I imagine it's too late to change this now.

On the other hand, what if `setState()` did return a Promise too?

Of course, what would that mean?

What if the Promise was fulfilled when any binding was completed and any resulting DOM mutations had occurred? Then we could still allow the current usage, and make sense out of the Promise version too.

@samouri , I'm sure you can improve this proposal...

**Comment by samouri** [comment]
I don't think we can make `getState` synchronous without blocking the event loop. `await` should make it easy for people to code as if it were though.

I agree that `setState` could return a Promise. Normally when you call `setState` there are two possible flows. One is to just set the state variable but skip dom mutations, and the other is to also update the dom. 

I'd think the Promise should fulfill when its completed doing whatever work it was going to do. If devs need access to whether or not we attempted to update the dom (called eval) we can add that as a resolved value.

**Comment by morsssss** [comment]
That's a nice idea. I like that!

This would add new functionality and improve consistency without breaking existing code.

**Comment by samouri** [comment]
As it turns out, Will was way ahead of us here :): https://github.com/ampproject/worker-dom/commit/1b15ccf3b2277dffd5a09cf82bc0c0c893505069#diff-adc7fa2d0014f654541b0fa9eaad6befR44

**Comment by morsssss** [comment]
:) I feel better when I see someone like Will also ending up not getting to all his TODOs... better to launch the MVP!
